MagazineLayout/LayoutCore/SectionModel.swift

```diff
@@ -305,22 +306,23 @@ struct SectionModel {
 
   // MARK: Private
 
+  private var numberOfRows: Int
   private var itemModels: [ItemModel]
   private var metrics: MagazineLayoutSectionMetrics
   private var calculatedHeight: CGFloat
 
-  private var indexOfFirstInvalidatedRow: Int?
+  private var indexOfFirstInvalidatedRow: Int? {
+    didSet {
+      guard indexOfFirstInvalidatedRow != nil else { return }
+      applyRowOffsetsIfNecessary()
+    }
+  }
```

Comment:
A future refactor could encapsulate all of this row offset / row <-> element mapping / row invalidation logic in a single object, which would be nicer than having multiple private properties in the section model to track all of this row-related state.